### DISCUSSION TITLE
strange error happen on firefox only/MOZILLA_PKIX_ERROR_REQUIRED_TLS_FEATURE_MISSING

### DISCUSSION THREAD

**Discussion by Giuseppe-TD** [ZoeyVid/NPMplus#1249] [Q&A] [answered]
in the latests updates i see a strange  error happening only in firefox:
MOZILLA_PKIX_ERROR_REQUIRED_TLS_FEATURE_MISSING
a function tls mandatory is not avaiable (or something similar, i have traslated)

but if i refresh the page the website load normally
(i  have many websites and all do the same)
i tried to renew the certificates yesterday, but no luck, after sometimes the error appears again, and this morning the errors agains
no logs on nginx, so i assume is something regarding firefox directly
on chromium browsers no errors

**Comment by Zoey2936** [comment]
(I thought this error would be fixed with ocsp fetcher...)
So you are accessing NPMplus over a private Network?

**Comment by Zoey2936** [comment]
either directly using a private ip/dns rewrite to private ip or using a public ip but you use the same public ip as NPMplus

**Comment by Giuseppe-TD** [comment]
yes i try to access to the websites on the same network as npmplus, so this should be the cause? in that case, no problem, I thought the problem is for everyone on firefox :) 

so you're planning to fix ocsp fetcher?

**Comment by Zoey2936** [comment]
the problem has everone using firefox and accessing NPMplus over a private ip (or the same public ip)

**Comment by Zoey2936** [comment]
the problem can be fixed using two ways: disabling ocsp fetching in firefox settings or disabling must-staple in NPMplus

**Comment by Zoey2936** [comment]
I will add an env/ui option to support disabling must staple

**Comment by Giuseppe-TD** [comment]
oh ok.. good, will wait for fix :D thank you

**Comment by Giuseppe-TD** [comment]
> the problem can be fixed using two ways: disabling ocsp fetching in firefox settings or disabling must-staple in NPMplus

disabling must-staple in npmplus can be dangerous? or a security risk?
because if only I have problem on my local network .. no problem :D

**Comment by Zoey2936** [comment]
(simplified) So in general, OCSP is used to check if a cert was revoked. must-staple enforces this check, so it needs to be checked otherwise it fails. There are two ways for this check: 1. the client asks the CA if the cert is valid (only done in firefox) or 2. stapling, which is done by NPMplus. Stapling is like a second short live cert in addition to the normal cert. Stapling works good and is no problem with must-staple and local ips. The client-side firefox check always fails on local ips, but only throws and error with must-staple enabled. 

Disabling this check in firefox (chromium has this always disabled) has pros and cons:
pro disabled in firefox:
- privacy: CA is not always asked about if a cert is valid (it knows your ip and cert you want to validate)
- makes must-staple work with local ips

cons disabled in firefox:
- firefox may not detect (compromised) revoked certs if they don't use must-staple, it may detect it since firefox checks them also against a local (always oudated) database with some (never all) revoked certs


same for must-staple:
pros:
- enforces validation using stapling

cons:
- breaks firefox on local ips


as you see this is not a yes or no question

**Comment by Zoey2936** [comment]
to talk into the future:
Lets encrypt wants to stop supporting ocsp at all (including server side stapling), because the client requests (like firefox does) are very expensive for them. They want to switch to these outdated local databases included in browsers.

**Comment by Giuseppe-TD** [comment]
oh .. i see.. so as a web dev, network admin, and so on, i will let everything as now, if the only "problem" i have is that on local i have this issue is not important, what i was worried is that everyone outside could have a problem (cause i host 20+ websites on my server), but if only occours on local, then is ok :D, i can live with that

**Comment by Giuseppe-TD** [comment]
> to talk into the future: Lets encrypt wants to stop supporting ocsp at all (including server side stapling), because the client requests (like firefox does) are very expensive for them. They want to switch to these outdated local databases included in browsers.

i think that firefox should disable this "feature" on his browser, 'cause let's encrypt is used everywhere on internet, not only on my websites :D

**Comment by Zoey2936** [comment]
To reduce the costs letsencrypt has? Possible, but I think to late and also it has some advantages. When they fully drop it will be like must staple is disabled and firefox check also

**Comment by Giuseppe-TD** [comment]
let's see in the future, anyway thank you for your help and explanation. if the problem remains on local, also in the future, no problems then :D

**Comment by Giuseppe-TD** [comment]
hi again, i would ask you a question, i receive this error not only in the local network but also when i'm home (same notebook) everytime i restart firefox, so, you told that you will add an env variable/ui button to disable must staple, when this will be ready? 
and another thing, could happen on the same notebook cause i use it in the local network then at home? (sort of certs cache) or it's indipendent?

**Comment by Zoey2936** [comment]
does the external ip also differ?

**Comment by Giuseppe-TD** [comment]
yes the external ip from my office and house are different

**Comment by Zoey2936** [comment]
and you don't use a vpn or something like that to open the page?

**Comment by Giuseppe-TD** [comment]
no, or .. not always sometimes i need a vpn to access my office folders, beside that everytime even without vpn i still have the warning, my sister in his house too (different ip) and she don't use vpn

**Comment by Zoey2936** [comment]
ok thats interesting, is there something written in the docker logs?

**Comment by Giuseppe-TD** [comment]
nope.. nothing in the npm plus docker logs ( i see only crowdsec bans, not to my ip )

**Comment by Zoey2936** [comment]
I will add an env option

**Comment by Giuseppe-TD** [comment]
Thank you 😁

**Comment by Smallinger** [comment]
Hey, I got this error today as well. I switched from Chrome to Firefox, and I hope I can manage it better with your environment option.

I tried different IP addresses, even used my phone without Wi-Fi and the Firefox app, but I’m still getting the error there as well. So, unfortunately, it’s definitely an issue with NPMPlus or Firefox. Unfortunately, there’s not much to see in the container logs either.

What I don’t understand is that if I refresh the website 2-3 times, I suddenly get access.

**Comment by Giuseppe-TD** [comment]
> Hey, I got this error today as well. I switched from Chrome to Firefox, and I hope I can manage it better with your environment option.
> 
> I tried different IP addresses, even used my phone without Wi-Fi and the Firefox app, but I’m still getting the error there as well. So, unfortunately, it’s definitely an issue with NPMPlus or Firefox. Unfortunately, there’s not much to see in the container logs either.
> 
> What I don’t understand is that if I refresh the website 2-3 times, I suddenly get access.

yes this is the exactly issue i get, Zoey will add and env option to disable must-staple so firefox will work :D 

i hope soon :D

**Comment by Zoey2936** [comment, marked as answer]
The env option is done and is released

**Reply by Felitendo** [reply]
Nice, thank you :)

**Comment by Giuseppe-TD** [comment]
i tried to enable the variable now but don't work and recreate all the certs, but as i read in you comment, this won't work after the update right?

**Reply by Zoey2936** [reply]
its not released yeet, but it will be released today

**Reply by Giuseppe-TD** [reply]
ok thanks

**Comment by MaximumFish** [comment]
Quick question: I assume that in order to recreate the cert I need to fully delete it and add it back in again? Will doing this mean that I need to go through all 28 of my hosts and add the new cert?

Perhaps there's a way to recreate it from the command line?

**Reply by Zoey2936** [reply]
> I assume that in order to recreate the cert I need to fully delete it and add it back in again?

yes

> Will doing this mean that I need to go through all 28 of my hosts and add the new cert?

yes

> Perhaps there's a way to recreate it from the command line?

no

**Reply by Giuseppe-TD** [reply]
> Quick question: I assume that in order to recreate the cert I need to fully delete it and add it back in again? Will doing this mean that I need to go through all 28 of my hosts and add the new cert?
> 
> Perhaps there's a way to recreate it from the command line?

i did for my 45 hosts too :D

**Reply by MaximumFish** [reply]
Edit: Read Zoey2936's comment below mine and be aware of the caveat if you're going to use this shortcut! You're probably better off just doing it the long way.

For anyone else coming to this thread, I managed to do it the easy way by:

1. Click the 3 dots next to your existing certificate. It'll say something like `Certificate #1`. Make a note of the number, we'll call it X.
2. Delete and recreate the cert.
3. Repeat step 1. We'll call the new number Y.
4. SSH onto your host box and enter your NPM proxy config directory. By default this is `/opt/npm/nginx/proxy_host`
5. run `sudo grep -l 'npm-X' . | sudo xargs sed -i 's/npm-X/npm-Y/g'` replacing X (twice) and Y as needed.
6. Restart npmplus.

**Reply by Zoey2936** [reply]
this will make no changes to the database, so no recommendation from me, if you resave a host without changing the cert it will use the old again. Also your guide will not change redirection or dead hosts

**Reply by MaximumFish** [reply]
I only use proxy hosts, but you raise good points, I will amend the comment.
